sparse/core.py

```diff
@@ -356,21 +374,20 @@ def reduction(self, method, axis=None, keepdims=False, dtype=None):
         return result
 
     def sum(self, axis=None, keepdims=False, dtype=None, out=None):
-        return self.reduction('sum', axis=axis, keepdims=keepdims, dtype=dtype)
+        assert out is None
+        return self.reduce(np.add, axis=axis, keepdims=keepdims, dtype=dtype)
 
     def max(self, axis=None, keepdims=False, out=None):
-        x = self.reduction('max', axis=axis, keepdims=keepdims)
-        x_zero = _zero_of_dtype(x.dtype)
-
-        # TODO: verify that there are some missing elements in each entry
-        if isinstance(x, COO):
-            x.data[x.data < x_zero] = x_zero
-            return x
-        elif isinstance(x, np.ndarray):
-            x[x < x_zero] = x_zero
-            return x
-        else:
-            return np.max(x, 0)
+        assert out is None
+        return self.reduce(np.maximum, axis=axis, keepdims=keepdims)
+
+    def min(self, axis=None, keepdims=False, out=None):
+        assert out is None
+        return self.reduce(np.minimum, axis=axis, keepdims=keepdims)
+
+    def prod(self, axis=None, keepdims=False, dtype=None, out=None):
+        assert out is None
+        return self.reduce(np.multiply, axis=axis, keepdims=keepdims, dtype=dtype)
```

Comment:
I'm curious, have you benchmarked the difference in performance here?

Reply:
Good point, I  just thought of that.
```
import numpy as np, sparse
from sparse.tests.test_core import random_x
x = random_x((10 ** 3, 10 ** 4))
xs = sparse.COO.from_numpy(x)
%timeit xs.sum(axis=1)
```

```
New: 1.31 ms ± 231 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Old: 11.8 ms ± 39.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

I certainly didn't expect that! I expected it to be marginally faster, not this much. With `axis=0`, though:
```
New: 83 ms ± 645 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Old: 9.8 ms ± 116 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

That's sad. But until `scipy.sparse` implements `__array_ufunc__`, we're stuck with this.

Reply:
> Edit: Most of the time seems to be taken in sort_indices in the second case, as expected.

Yeah, I was just looking at this as well.  It looks like it remains slow on repeated computation as well.  I suspect that sortedness does not transfer through the transpose operation.  I wonder if there is an efficient way to sort more efficiently in this case.  (this isn't important to do now, I'm just thinking out loud.)

Reply:
Of course, integer sorting can be made `O(n)`, while still keeping it stable. But implementing it will require Cython/Numba, at the very least, as Numpy doesn't have it.

Reply:
See, for example, [this](https://stackoverflow.com/questions/35317442/how-to-sort-an-array-of-integers-faster-than-quicksort)

Reply:
xref https://github.com/numpy/numpy/issues/6050

Reply:
We can also cache transpose operations. However, it doesn't make sense that the same sum will be done twice, so...

Reply:
Another thing we can do (which is a completely huge project on its own) is to switch to KD trees - they're always sorted, we never have to sort. `__getitem__` will also be faster. However, the cost is slower indexing.